File: SmartRoutine/backend/src/main/java/service/AlimentoService.java
package service;

import com.google.gson.Gson;
import dao.AlimentoDAO;
import model.Alimento;
import spark.Request;
import spark.Response;

public class AlimentoService {

    private AlimentoDAO alimentoDAO;
    private Gson gson;

    public AlimentoService() {
        this.alimentoDAO = new AlimentoDAO();
        this.gson = new Gson();
    }

    /**
     * GET /alimento
     * Lista todos os alimentos
     */
    public Object getAll(Request request, Response response) {
        response.type("application/json");

        try {
            return gson.toJson(alimentoDAO.getAll());
        } catch (Exception e) {
            response.status(500);
            return gson.toJson(new ErrorResponse("Erro ao listar alimentos: " + e.getMessage()));
        }
    }

    /**
     * GET /alimento/:id
     * Busca alimento por ID
     */
    public Object get(Request request, Response response) {
        response.type("application/json");

        try {
            int id = Integer.parseInt(request.params(":id"));
            Alimento alimento = alimentoDAO.get(id);

            if (alimento != null) {
                response.status(200);
                return gson.toJson(alimento);
            } else {
                response.status(404);
                return gson.toJson(new ErrorResponse("Alimento não encontrado"));
            }
        } catch (NumberFormatException e) {
            response.status(400);
            return gson.toJson(new ErrorResponse("ID inválido"));
        } catch (Exception e) {
            response.status(500);
            return gson.toJson(new ErrorResponse("Erro ao buscar alimento: " + e.getMessage()));
        }
    }

    /**
     * GET /alimento/categoria/:categoria
     * Lista alimentos por categoria
     */
    public Object getByCategoria(Request request, Response response) {
        response.type("application/json");

        try {
            String categoria = request.params(":categoria");
            return gson.toJson(alimentoDAO.getByCategoria(categoria));
        } catch (Exception e) {
            response.status(500);
            return gson.toJson(new ErrorResponse("Erro ao buscar alimentos: " + e.getMessage()));
        }
    }

    /**
     * GET /alimento/search?q=nome
     * Busca alimentos por nome
     */
    public Object search(Request request, Response response) {
        response.type("application/json");

        try {
            String query = request.queryParams("q");
            if (query == null || query.trim().isEmpty()) {
                response.status(400);
                return gson.toJson(new ErrorResponse("Parâmetro 'q' é obrigatório"));
            }
            return gson.toJson(alimentoDAO.searchByNome(query));
        } catch (Exception e) {
            response.status(500);
            return gson.toJson(new ErrorResponse("Erro ao buscar alimentos: " + e.getMessage()));
        }
    }

    /**
     * GET /alimento/categorias
     * Lista todas as categorias
     */
    public Object getCategorias(Request request, Response response) {
        response.type("application/json");

        try {
            return gson.toJson(alimentoDAO.getCategorias());
        } catch (Exception e) {
            response.status(500);
            return gson.toJson(new ErrorResponse("Erro ao listar categorias: " + e.getMessage()));
        }
    }

    /**
     * POST /alimento
     * Insere um novo alimento
     */
    public Object insert(Request request, Response response) {
        response.type("application/json");

        try {
            Alimento alimento = gson.fromJson(request.body(), Alimento.class);

            // Validações
            if (alimento.getNome() == null || alimento.getNome().trim().isEmpty()) {
                response.status(400);
                return gson.toJson(new ErrorResponse("Nome é obrigatório"));
            }

            if (alimentoDAO.insert(alimento)) {
                response.status(201);
                return gson.toJson(new SuccessResponse("Alimento cadastrado com sucesso"));
            } else {
                response.status(500);
                return gson.toJson(new ErrorResponse("Erro ao cadastrar alimento"));
            }
        } catch (Exception e) {
            response.status(400);
            return gson.toJson(new ErrorResponse("Dados inválidos: " + e.getMessage()));
        }
    }

    /**
     * PUT /alimento/:id
     * Atualiza um alimento existente
     */
    public Object update(Request request, Response response) {
        response.type("application/json");

        try {
            int id = Integer.parseInt(request.params(":id"));
            Alimento alimentoExistente = alimentoDAO.get(id);

            if (alimentoExistente == null) {
                response.status(404);
                return gson.toJson(new ErrorResponse("Alimento não encontrado"));
            }

            Alimento alimento = gson.fromJson(request.body(), Alimento.class);
            alimento.setId(id);

            // Validações
            if (alimento.getNome() == null || alimento.getNome().trim().isEmpty()) {
                response.status(400);
                return gson.toJson(new ErrorResponse("Nome é obrigatório"));
            }

            if (alimentoDAO.update(alimento)) {
                response.status(200);
                return gson.toJson(new SuccessResponse("Alimento atualizado com sucesso"));
            } else {
                response.status(500);
                return gson.toJson(new ErrorResponse("Erro ao atualizar alimento"));
            }
        } catch (NumberFormatException e) {
            response.status(400);
            return gson.toJson(new ErrorResponse("ID inválido"));
        } catch (Exception e) {
            response.status(400);
            return gson.toJson(new ErrorResponse("Dados inválidos: " + e.getMessage()));
        }
    }

    /**
     * DELETE /alimento/:id
     * Remove um alimento
     */
    public Object delete(Request request, Response response) {
        response.type("application/json");

        try {
            int id = Integer.parseInt(request.params(":id"));
            Alimento alimento = alimentoDAO.get(id);

            if (alimento == null) {
                response.status(404);
                return gson.toJson(new ErrorResponse("Alimento não encontrado"));
            }

            if (alimentoDAO.delete(id)) {
                response.status(200);
                return gson.toJson(new SuccessResponse("Alimento removido com sucesso"));
            } else {
                response.status(500);
                return gson.toJson(new ErrorResponse("Erro ao remover alimento"));
            }
        } catch (NumberFormatException e) {
            response.status(400);
            return gson.toJson(new ErrorResponse("ID inválido"));
        } catch (Exception e) {
            response.status(500);
            return gson.toJson(new ErrorResponse("Erro ao deletar alimento: " + e.getMessage()));
        }
    }

    // Classes auxiliares
    private static class ErrorResponse {
        private String error;
        public ErrorResponse(String error) {
            this.error = error;
        }
    }

    private static class SuccessResponse {
        private String message;
        public SuccessResponse(String message) {
            this.message = message;
        }
    }
}

File: SmartRoutine/frontend/src/services/alimentoService.ts
import { api } from './api';
import { handleApiError } from './errorHandler';
import { Alimento } from '../types';

export const alimentoService = {
  /**
   * Lista todos os alimentos
   */
  async getAll(): Promise<Alimento[]> {
    try {
      const response = await api.get<Alimento[]>('/alimento');
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Busca alimento por ID
   */
  async getById(id: number): Promise<Alimento> {
    try {
      const response = await api.get<Alimento>(`/alimento/${id}`);
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Lista alimentos por categoria
   */
  async getByCategoria(categoria: string): Promise<Alimento[]> {
    try {
      const response = await api.get<Alimento[]>(`/alimento/categoria/${categoria}`);
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Busca alimentos por nome
   */
  async search(query: string): Promise<Alimento[]> {
    try {
      const response = await api.get<Alimento[]>('/alimento/search', {
        params: { q: query },
      });
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Lista todas as categorias disponíveis
   */
  async getCategorias(): Promise<string[]> {
    try {
      const response = await api.get<string[]>('/alimento/categorias');
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Cria novo alimento no catálogo
   */
  async create(alimento: Omit<Alimento, 'id'>): Promise<{ message: string }> {
    try {
      const response = await api.post('/alimento', alimento);
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Atualiza alimento existente
   */
  async update(id: number, alimento: Partial<Alimento>): Promise<{ message: string }> {
    try {
      const response = await api.put(`/alimento/${id}`, alimento);
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Deleta alimento do catálogo
   */
  async delete(id: number): Promise<{ message: string }> {
    try {
      const response = await api.delete(`/alimento/${id}`);
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },

  /**
   * Busca alimentos com paginação (se implementado no backend)
   */
  async getPaginated(page: number = 1, limit: number = 20): Promise<{
    alimentos: Alimento[];
    total: number;
    page: number;
    totalPages: number;
  }> {
    try {
      const response = await api.get('/alimento/paginated', {
        params: { page, limit },
      });
      return response.data;
    } catch (error) {
      handleApiError(error);
    }
  },
};